mayastor/src/subsys/mbus/registration.rs

```diff
@@ -0,0 +1,203 @@
+        MESSAGE_BUS_REG.get_or_init(|| Registration::new(node, grpc_endpoint));
+    }
+
+    /// terminate and re-register
+    pub(super) fn fini(&self) {
+        self.fini_chan.close();
+    }
+
+    pub(super) fn get() -> &'static Registration {
+        MESSAGE_BUS_REG.get().unwrap()
+    }
+
+    /// runner responsible for registering and
+    /// de-registering the mayastor instance on shutdown
+    pub async fn run() -> Result<(), ()> {
+        if let Some(registration) = MESSAGE_BUS_REG.get() {
+            registration.clone().run_loop().await;
+        }
+        Ok(())
+    }
+
+    fn new(node: &str, grpc_endpoint: &str) -> Registration {
+        let (msg_sender, msg_receiver) = smol::channel::unbounded::<()>();
+        let config = Configuration {
+            node: node.to_owned(),
+            grpc_endpoint: grpc_endpoint.to_owned(),
+            hb_interval: match env::var("MAYASTOR_HB_INTERVAL")
+                .map(|v| v.parse::<u64>())
+            {
+                Ok(Ok(num)) => Duration::from_secs(num),
+                _ => HB_INTERVAL,
+            },
+        };
+        Self {
+            config,
+            rcv_chan: msg_receiver,
+            fini_chan: msg_sender,
+        }
+    }
+}
+
+impl Registration {
+    /// Connect to the server and start emitting periodic register
+    /// messages.
+    /// Runs until the sender side of the message channel is closed
+    pub async fn run_loop(&mut self) {
+        info!(
+            "Registering '{}' and grpc server {} ...",
+            self.config.node, self.config.grpc_endpoint
+        );
+        loop {
+            if let Err(err) = self.register().await {
+                error!("Registration failed: {:?}", err);
+            };
+
+            select! {
+                _ = smol::Timer::after(self.config.hb_interval).fuse() => continue,
+                msg = self.rcv_chan.next().fuse() => {
+                    match msg {
+                        Some(_) => log::info!("Messages have not been implemented yet"),
+                        _ => {
+                            log::info!("Terminating the registration handler");
+                            break;
+                        }
+                    }
+                }
+            };
+        }
+        if let Err(err) = self.deregister().await {
+            error!("Deregistration failed: {:?}", err);
+        };
+    }
+
+    /// Send a register message to the MessageBus.
+    async fn register(&self) -> Result<(), Error> {
+        let payload = RegisterArgs {
+            id: self.config.node.clone(),
+            grpc_endpoint: self.config.grpc_endpoint.clone(),
+        };
+        super::message_bus()
+            .publish("register", serde_json::to_vec(&payload).unwrap())
+            .await
+            .map_err(|cause| Error::QueueRegister {
+                cause,
+            })?;
+
+        // Note that the message was only queued and we don't know if it was
```

Comment:
having a way to figure out that we are registered, seems to make a whole lot of sense to me. In particular, during production issues where there seems to be a disconnect somehow somewhere. 

If we set some variable somehow and make that readable/gettable trough RPC we can then ensure: "yes -- this instance is regisitered" -- having this in the logs then would also help.

Reply:
We can do that via the request I think. We'll have to figure out how to implement that, probably using request/reply with a timeout retry loop.
The way it works is you send to channel X, message Y and wait on channel Z for the reply. The other side then receives Y on channel X and then replies to channel Z. 
We have to retry in case the other side was not connected when the message was sent of if it crashed after receiving it and before replying.